File: cpp_merge/packages/integration/SWEagent/src/run/remove-unfinished.cpp
#include "remove-unfinished.h"

std::shared_ptr<Promise<void>> removeUnfinished(string baseDir, boolean dryRun)
{
    auto basePath = path->resolve(baseDir);
    auto toRemove = array<string>();
    auto directories = fs::readdirSync(basePath);
    auto& __array627_1915 = directories;
    for (auto __indx627_1915 = 0_N; __indx627_1915 < __array627_1915->get_length(); __indx627_1915++)
    {
        auto& dirName = const_(__array627_1915)[__indx627_1915];
        {
            shared directory = path->join(basePath, dirName);
            if (!fs::statSync(directory)->isDirectory()) {
                continue;
            }
            if (!dirName->includes(std:("__"))) {
                continue;
            }
            auto trajs = fs::readdirSync(directory)->filter([=](auto file) mutable
            {
                return file->endsWith(std:(".traj"));
            }
            )->map([=](auto file) mutable
            {
                return path->join(directory, file);
            }
            );
            if (trajs->get_length() == 0) {
                logger->info(std:("No trajectories found in ") + directory + string_empty);
                continue;
            }
            if (trajs->get_length() > 1) {
                logger->warn(std:("Found multiple trajectories in ") + directory + std:(". Skipping."));
                continue;
            }
            try
            {
                auto traj = loadFile(const_(trajs)[0]);
                if (OR((!traj), (type_of(traj) != std:("object")))) {
                    logger->warn(std:("Invalid trajectory format in ") + const_(trajs)[0] + std:(". Adding to remove list."));
                    toRemove->push(directory);
                    continue;
                }
                auto submission = OR(((as<std::shared_ptr<TrajectoryFile>>(traj))->info["submission"]), (nullptr));
                if (submission == nullptr) {
                    logger->warn(std:("No submission found in ") + directory + std:(". Adding to remove list."));
                    toRemove->push(directory);
                    continue;
                }
            }
            catch (const any& error)
            {
                logger->warn(std:("Error loading trajectory ") + const_(trajs)[0] + std:(": ") + error + std:(". Adding to remove list."));
                toRemove->push(directory);
                continue;
            }
        }
    }
    if (dryRun) {
        logger->info(std:("Would remove ") + toRemove->get_length() + std:(" unfinished trajectories."));
        for (auto& directory : toRemove)
        {
            logger->info(directory);
        }
    } else {
        for (auto& directory : toRemove)
        {
            logger->info(std:("Removing ") + directory + string_empty);
            fs::rmSync(directory, object{
                object::pair{std:("recursive"), true}, 
                object::pair{std:("force"), true}
            });
        }
    }
    return std::shared_ptr<Promise<void>>();
};


std::shared_ptr<AgentLogger> logger = getLogger(std:("remove_unfinished"));

void Main(void)
{
}

MAIN

File: SWEagent/src/run/remove-unfinished.ts
/**
 * Remove unfinished trajectories
 */

import * as path from 'path';
import * as fs from 'fs';
import { loadFile } from '../utils/files';
import { getLogger } from '../utils/log';

const logger = getLogger('remove_unfinished');

interface TrajectoryFile {
  info?: {
    submission?: string | null;
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

/**
 * Remove unfinished trajectories
 */
export async function removeUnfinished(baseDir: string, dryRun: boolean = true): Promise<void> {
  const basePath = path.resolve(baseDir);
  const toRemove: string[] = [];

  const directories = fs.readdirSync(basePath);

  for (const dirName of directories) {
    const directory = path.join(basePath, dirName);

    if (!fs.statSync(directory).isDirectory()) {
      continue;
    }

    if (!dirName.includes('__')) {
      continue;
    }

    const trajs = fs
      .readdirSync(directory)
      .filter((file) => file.endsWith('.traj'))
      .map((file) => path.join(directory, file));

    if (trajs.length === 0) {
      logger.info(`No trajectories found in ${directory}`);
      continue;
    }

    if (trajs.length > 1) {
      logger.warn(`Found multiple trajectories in ${directory}. Skipping.`);
      continue;
    }

    try {
      const traj = loadFile(trajs[0]);

      if (!traj || typeof traj !== 'object') {
        logger.warn(`Invalid trajectory format in ${trajs[0]}. Adding to remove list.`);
        toRemove.push(directory);
        continue;
      }

      const submission = (traj as TrajectoryFile).info?.submission || null;

      if (submission === null) {
        logger.warn(`No submission found in ${directory}. Adding to remove list.`);
        toRemove.push(directory);
        continue;
      }
    } catch (error) {
      logger.warn(`Error loading trajectory ${trajs[0]}: ${error}. Adding to remove list.`);
      toRemove.push(directory);
      continue;
    }
  }

  if (dryRun) {
    logger.info(`Would remove ${toRemove.length} unfinished trajectories.`);
    for (const directory of toRemove) {
      logger.info(directory);
    }
  } else {
    for (const directory of toRemove) {
      logger.info(`Removing ${directory}`);
      fs.rmSync(directory, { recursive: true, force: true });
    }
  }
}
